### The following processes will be conducted in this data wrangling section:

### Reading the raw data from a CSV file

### Display data samples to understand its structure

### Data Cleaning: Identify and handle missing values

### Feature Engineering: Create new features based on existing ones (e.g., calculate mean income per zip code)

### Data Filtering: Filter out properties that are not relevant to the analysis (e.g., outliers or properties outside Austin city limits)

### # Feature Engineering: Create new features based on existing ones (e.g., calculate price per square foot)

### Data Aggregation: Calculate average price per zip code

### Data Merging: Combining housing data with demographic data

### Data Visualization: Visualize the data to gain insights and identify trends

### After data wrangling, the cleaned and transformed data can be used for further analysis and modeling.


In [1]:
# First, import the relevant modules and packages
import pandas as pd
import numpy as np
#from matplotlib import pyplot
import matplotlib.pyplot as plt

In [2]:
# Import Austin_TX_House_Listings_data_v1.csv & us_income_zipcode.csv
df1 = pd.read_csv('../01_raw_data/AustinTXHouseListingsDataV1.csv', low_memory=False)

In [3]:
# Displaying the first few rows of the raw data to understand its structure
df1.head()

Unnamed: 0.1  Unnamed: 0 address_city          address_neighborhood  \
0             0          15       Austin                           NaN   
1             1          16       Austin                           NaN   
2             2          17       Austin                           NaN   
3             3          18       Austin                           NaN   
4             4          19       Austin  Steiner Ranch-Lakewood Hills   

  address_state    address_streetAddress address_subdivision  address_zipcode  \
0            TX     12801 Wooded Lake Ct                 NaN            78732   
1            TX   904 Lakewood Hills Ter                 NaN            78732   
2            TX        13701 Montview Dr                 NaN            78732   
3            TX   700 Lakewood Hills Ter                 NaN            78732   
4            TX  1008 Lakewood Hills Ter                 NaN            78732   

   bathrooms  bedrooms  ...  latest_salemonth latest_saleyear  \
0        5.0       4.0  ...                 2            2018   
1        5.0       5.0  ...                 8            2020   
2        3.5       NaN  ...                 3            2018   
3        4.0       4.0  ...                 6            2018   
4        4.0       4.0  ...                 8            2019   

  latest_price_source  numOfAccessibilityFeatures  numOfAppliances  numPhotos  \
0      Agent Provided                           0                3         39   
1     Broker Provided                           0                3         58   
2      Agent Provided                           0                0          1   
3      Agent Provided                           0                1         40   
4     Broker Provided                           0                4         40   

   numPriceChanges                                              photo  \
0                1  https://photos.zillowstatic.com/fp/911f9a59acb...   
1                2  https://photos.zillowstatic.com/fp/c6346c4a39d...   
2                1  https://maps.googleapis.com/maps/api/streetvie...   
3                4  https://photos.zillowstatic.com/fp/805b5a6b748...   
4                7  https://photos.zillowstatic.com/fp/a09cfad45a9...   

                                           imagepath  \
0  /content/gdrive/MyDrive/zillow-images/70352485...   
1  /content/gdrive/MyDrive/zillow-images/70352465...   
2  /content/gdrive/MyDrive/zillow-images/83823478...   
3  /content/gdrive/MyDrive/zillow-images/70352478...   
4  /content/gdrive/MyDrive/zillow-images/70352461...   

                                           cleanpath  
0  70352485_911f9a59acb6fe5fd909538a685e2a0c-p_f.jpg  
1  70352465_c6346c4a39d1f87578ee1529329d9296-p_f.jpg  
2  83823478_streetviewlocation13701MontviewDr2CAu...  
3  70352478_805b5a6b748b1d9cffeb93a4c41fedb0-p_f.jpg  
4  70352461_a09cfad45a949f7195c9540224f51e49-p_f.jpg  

[5 rows x 736 columns]

In [4]:
df1.shape

(16482, 736)

It looks like our data has more features (columns) than we need

In [5]:
#Analyze the summary of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16482 entries, 0 to 16481
Columns: 736 entries, Unnamed: 0.1 to cleanpath
dtypes: bool(13), float64(227), int64(10), object(486)
memory usage: 91.1+ MB


In [6]:
df1.columns

Index(['Unnamed: 0.1', 'Unnamed: 0', 'address_city', 'address_neighborhood',
       'address_state', 'address_streetAddress', 'address_subdivision',
       'address_zipcode', 'bathrooms', 'bedrooms',
       ...
       'latest_salemonth', 'latest_saleyear', 'latest_price_source',
       'numOfAccessibilityFeatures', 'numOfAppliances', 'numPhotos',
       'numPriceChanges', 'photo', 'imagepath', 'cleanpath'],
      dtype='object', length=736)

In [7]:
# Analyze how the dependent variable, 'latest_price', is correlated with the rest of independent variables
dict(df1.corr()['latest_price'].abs().sort_values(ascending = False))

C:\Users\famma\AppData\Local\Temp\ipykernel_34148\1946049958.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  dict(df1.corr()['latest_price'].abs().sort_values(ascending = False))


{'latest_price': 1.0,
 'taxAssessedValue': 0.8628559938898727,
 'taxAnnualAmount': 0.8156128587000064,
 'bathrooms': 0.551445307264037,
 'resoFactsStats_bathrooms': 0.47586093567052096,
 'livingArea': 0.3919678078297034,
 'resoFactsStats_bathroomsFull': 0.3854093485257164,
 'bedrooms': 0.3236994120035267,
 'resoFactsStats_bedrooms': 0.29769330047900794,
 'schools_2_rating': 0.2879634712029946,
 'schools_1_studentsPerTeacher': 0.2846390962391095,
 'schools_1_rating': 0.24097254565070483,
 'schools_0_rating': 0.23987455502465402,
 'schools_1_size': 0.23598605610932682,
 'resoFactsStats_bathroomsHalf': 0.22839099263704998,
 'stories': 0.21967042532231817,
 'numPhotos': 0.181905441001118,
 'longitude': 0.18157016525279043,
 'hasSpa': 0.17628657766491157,
 'coveredSpaces': 0.17196269771541464,
 'parking': 0.16639348500886164,
 'schools_2_studentsPerTeacher': 0.16305886359458596,
 'garageSpaces': 0.16297483664756376,
 'schools_0_distance': 0.15324071928582075,
 'address_zipcode': 0.152604521

In [8]:
#Create a "corr_r" row to filter out less correlated features (columns)
df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)

C:\Users\famma\AppData\Local\Temp\ipykernel_34148\4094108136.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)
C:\Users\famma\AppData\Local\Temp\ipykernel_34148\4094108136.py:2: FutureWarning: Behavior when concatenating bool-dtype and numeric-dtype arrays is deprecated; in a future version these will cast to object dtype (instead of coercing bools to numeric values). To retain the old behavior, explicitly cast bool-dtype arrays to numeric dtype.
  df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)


In [9]:
# Check the new row
df1.tail()

Unnamed: 0.1    Unnamed: 0 address_city address_neighborhood  \
16478    16480.00000  22070.000000       Austin                  NaN   
16479    16481.00000  22072.000000       Austin                  NaN   
16480    16482.00000  22073.000000       Austin                  NaN   
16481    16483.00000  22074.000000       Austin                  NaN   
corr_r       0.12465      0.124756          NaN                  NaN   

       address_state address_streetAddress address_subdivision  \
16478             TX     6318 Clairmont Dr                 NaN   
16479             TX      6104 Abilene Trl                 NaN   
16480             TX      7702 Kincheon Ct                 NaN   
16481             TX         5300 Indio Cv                 NaN   
corr_r           NaN                   NaN                 NaN   

        address_zipcode  bathrooms  bedrooms  ...  latest_salemonth  \
16478      78749.000000   2.000000  4.000000  ...          1.000000   
16479      78749.000000   2.000000  3.000000  ...          1.000000   
16480      78749.000000   2.500000  3.000000  ...          1.000000   
16481      78745.000000        NaN       NaN  ...          1.000000   
corr_r         0.152605   0.551445  0.323699  ...          0.030104   

       latest_saleyear latest_price_source  numOfAccessibilityFeatures  \
16478      2018.000000      Agent Provided                    0.000000   
16479      2018.000000      Agent Provided                    0.000000   
16480      2018.000000      Agent Provided                    0.000000   
16481      2018.000000      Agent Provided                    0.000000   
corr_r        0.059022                 NaN                    0.015614   

        numOfAppliances  numPhotos  numPriceChanges  \
16478          3.000000  33.000000          1.00000   
16479          5.000000  25.000000          5.00000   
16480          1.000000  27.000000          1.00000   
16481          0.000000  11.000000          1.00000   
corr_r         0.048005   0.181905          0.09137   

                                                    photo  \
16478   https://photos.zillowstatic.com/fp/2f26fb52681...   
16479   https://photos.zillowstatic.com/fp/71a9662d303...   
16480   https://photos.zillowstatic.com/fp/f452bc0f4ba...   
16481   https://photos.zillowstatic.com/fp/eb612b3f7c8...   
corr_r                                                NaN   

                                                imagepath  \
16478   /content/gdrive/MyDrive/zillow-images/29487158...   
16479   /content/gdrive/MyDrive/zillow-images/29491564...   
16480   /content/gdrive/MyDrive/zillow-images/29486556...   
16481   /content/gdrive/MyDrive/zillow-images/58314089...   
corr_r                                                NaN   

                                                cleanpath  
16478   29487158_2f26fb52681290dd3e096cb85fc299ec-p_f.jpg  
16479   29491564_71a9662d3031a1834e29f578f5a92b71-p_f.jpg  
16480   29486556_f452bc0f4ba54fe11c89905773f38840-p_f.jpg  
16481   58314089_eb612b3f7c80976715d8a54e389b8e15-p_f.jpg  
corr_r                                                NaN  

[5 rows x 736 columns]

In [10]:
 df1.shape

(16483, 736)

In [11]:
# Exclude columns that have correlation values less than 0.05
df2 = df1.loc[:,df1.loc['corr_r'] >= 0.05]

In [12]:
df2.shape

(16483, 44)

The number of columns are reduced from 736 to 44 by dropping columns that have less than 0.05 correlation value 

In [13]:
df2.head()

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms  dateposted  \
0           0.0        15.0          78732.0        5.0       4.0         NaN   
1           1.0        16.0          78732.0        5.0       5.0         NaN   
2           2.0        17.0          78732.0        3.5       NaN         NaN   
3           3.0        18.0          78732.0        4.0       4.0         NaN   
4           4.0        19.0          78732.0        4.0       4.0         NaN   

    latitude  livingArea  longitude  propertyTaxRate  ...  \
0  30.354322      4060.0 -97.911278             1.98  ...   
1  30.355553      4558.0 -97.912544             1.98  ...   
2  30.382227      3402.0 -97.908333             1.98  ...   
3  30.352081      4749.0 -97.912048             1.98  ...   
4  30.356226      4867.0 -97.911697             1.98  ...   

   schools_1_studentsPerTeacher  schools_1_totalCount  schools_2_distance  \
0                          17.0                   1.0                 6.0   
1                          17.0                   1.0                 6.1   
2                          17.0                   1.0                 5.3   
3                          17.0                   1.0                 6.1   
4                          17.0                   1.0                 6.0   

   schools_2_rating  schools_2_studentsPerTeacher  schools_2_totalCount  \
0               7.0                          17.0                   1.0   
1               7.0                          17.0                   1.0   
2               7.0                          17.0                   1.0   
3               7.0                          17.0                   1.0   
4               7.0                          17.0                   1.0   

   latest_price  latest_saleyear  numPhotos numPriceChanges  
0      715000.0           2018.0       39.0             1.0  
1     1025000.0           2020.0       58.0             2.0  
2       59000.0           2018.0        1.0             1.0  
3      825000.0           2018.0       40.0             4.0  
4      849000.0           2019.0       40.0             7.0  

[5 rows x 44 columns]

In [14]:
# Create a row called 'null_sum_col' to drop columns that have more than 50% missing values
df2.loc['null_sum_col'] = df2.isnull().sum()

C:\Users\famma\AppData\Local\Temp\ipykernel_34148\3052481394.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2.loc['null_sum_col'] = df2.isnull().sum()


In [15]:
# Filter out rows that has more than 50% missing values
df3 = df2.loc[:, df2.loc['null_sum_col'] < 8242]

In [16]:
df3.shape

(16484, 37)

In [17]:
df3.head()

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms   latitude  \
0           0.0        15.0          78732.0        5.0       4.0  30.354322   
1           1.0        16.0          78732.0        5.0       5.0  30.355553   
2           2.0        17.0          78732.0        3.5       NaN  30.382227   
3           3.0        18.0          78732.0        4.0       4.0  30.352081   
4           4.0        19.0          78732.0        4.0       4.0  30.356226   

   livingArea  longitude  propertyTaxRate  resoFactsStats_bathrooms  ...  \
0      4060.0 -97.911278             1.98                       5.0  ...   
1      4558.0 -97.912544             1.98                       5.0  ...   
2      3402.0 -97.908333             1.98                       NaN  ...   
3      4749.0 -97.912048             1.98                       4.0  ...   
4      4867.0 -97.911697             1.98                       4.0  ...   

   schools_1_studentsPerTeacher  schools_1_totalCount  schools_2_distance  \
0                          17.0                   1.0                 6.0   
1                          17.0                   1.0                 6.1   
2                          17.0                   1.0                 5.3   
3                          17.0                   1.0                 6.1   
4                          17.0                   1.0                 6.0   

   schools_2_rating schools_2_studentsPerTeacher schools_2_totalCount  \
0               7.0                         17.0                  1.0   
1               7.0                         17.0                  1.0   
2               7.0                         17.0                  1.0   
3               7.0                         17.0                  1.0   
4               7.0                         17.0                  1.0   

  latest_price  latest_saleyear  numPhotos  numPriceChanges  
0     715000.0           2018.0       39.0              1.0  
1    1025000.0           2020.0       58.0              2.0  
2      59000.0           2018.0        1.0              1.0  
3     825000.0           2018.0       40.0              4.0  
4     849000.0           2019.0       40.0              7.0  

[5 rows x 37 columns]

The number of columns further decreased to 37

In [18]:
# Check how many rows have a total of more than 5 null values
df3[df3.isnull().sum(axis = 1) > 3]

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms  \
2               2.0        17.0          78732.0        3.5       NaN   
78             78.0       106.0          78732.0        NaN       NaN   
117           117.0       150.0          78732.0        NaN       NaN   
122           122.0       155.0          78732.0        NaN       NaN   
128           128.0       164.0          78734.0        NaN       NaN   
...             ...         ...              ...        ...       ...   
16360       16362.0     21631.0          78702.0        NaN       NaN   
16371       16373.0     21653.0          78702.0        NaN       NaN   
16382       16384.0     21667.0          78702.0        2.0       3.0   
16391       16393.0     21716.0          78737.0        3.0       4.0   
16417       16419.0     21857.0          78737.0        3.0       5.0   

        latitude  livingArea  longitude  propertyTaxRate  \
2      30.382227      3402.0 -97.908333             1.98   
78     30.382215      2512.0 -97.909950             1.98   
117    30.335987         NaN -97.923317             1.98   
122    30.329601     12306.0 -97.923370             1.98   
128    30.369040         NaN -97.915916             1.98   
...          ...         ...        ...              ...   
16360  30.270145         NaN -97.717087             1.98   
16371  30.268822         NaN -97.716232             1.98   
16382  30.268017      1260.0 -97.722214             1.98   
16391  30.163763      2582.0 -97.940430             2.01   
16417  30.198322      2747.0 -97.989227             2.01   

       resoFactsStats_bathrooms  ...  schools_1_studentsPerTeacher  \
2                           NaN  ...                          17.0   
78                          0.0  ...                          17.0   
117                         0.0  ...                          17.0   
122                         0.0  ...                          17.0   
128                         0.0  ...                          16.0   
...                         ...  ...                           ...   
16360                       0.0  ...                          15.0   
16371                       0.0  ...                          15.0   
16382                       2.0  ...                           NaN   
16391                       3.0  ...                           NaN   
16417                       3.0  ...                          16.0   

       schools_1_totalCount  schools_2_distance  schools_2_rating  \
2                       1.0                 5.3               7.0   
78                      1.0                 5.4               7.0   
117                     1.0                 7.2               7.0   
122                     1.0                 7.5               7.0   
128                     1.0                 4.2               7.0   
...                     ...                 ...               ...   
16360                   1.0                 2.3               4.0   
16371                   1.0                 2.3               4.0   
16382                   NaN                 NaN               NaN   
16391                   NaN                 NaN               NaN   
16417                   1.0                 NaN               NaN   

      schools_2_studentsPerTeacher schools_2_totalCount latest_price  \
2                             17.0                  1.0      59000.0   
78                            17.0                  1.0     100000.0   
117                           17.0                  1.0     850000.0   
122                           17.0                  1.0     164900.0   
128                           18.0                  1.0     129000.0   
...                            ...                  ...          ...   
16360                         11.0                  1.0     500000.0   
16371                         11.0                  1.0     300000.0   
16382                          NaN                  NaN     579000.0   
16391              

In [19]:
df3['null_sum_rows'] = df3.isnull().sum(axis = 1)

C:\Users\famma\AppData\Local\Temp\ipykernel_34148\365114040.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['null_sum_rows'] = df3.isnull().sum(axis = 1)


In [20]:
df3.head()

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms   latitude  \
0           0.0        15.0          78732.0        5.0       4.0  30.354322   
1           1.0        16.0          78732.0        5.0       5.0  30.355553   
2           2.0        17.0          78732.0        3.5       NaN  30.382227   
3           3.0        18.0          78732.0        4.0       4.0  30.352081   
4           4.0        19.0          78732.0        4.0       4.0  30.356226   

   livingArea  longitude  propertyTaxRate  resoFactsStats_bathrooms  ...  \
0      4060.0 -97.911278             1.98                       5.0  ...   
1      4558.0 -97.912544             1.98                       5.0  ...   
2      3402.0 -97.908333             1.98                       NaN  ...   
3      4749.0 -97.912048             1.98                       4.0  ...   
4      4867.0 -97.911697             1.98                       4.0  ...   

   schools_1_totalCount  schools_2_distance  schools_2_rating  \
0                   1.0                 6.0               7.0   
1                   1.0                 6.1               7.0   
2                   1.0                 5.3               7.0   
3                   1.0                 6.1               7.0   
4                   1.0                 6.0               7.0   

   schools_2_studentsPerTeacher schools_2_totalCount latest_price  \
0                          17.0                  1.0     715000.0   
1                          17.0                  1.0    1025000.0   
2                          17.0                  1.0      59000.0   
3                          17.0                  1.0     825000.0   
4                          17.0                  1.0     849000.0   

  latest_saleyear  numPhotos  numPriceChanges  null_sum_rows  
0          2018.0       39.0              1.0              0  
1          2020.0       58.0              2.0              0  
2          2018.0        1.0              1.0              5  
3          2018.0       40.0              4.0              0  
4          2019.0       40.0              7.0              0  

[5 rows x 38 columns]

In [21]:
df3.shape

(16484, 38)

In [22]:
df3 = df3[df3['null_sum_rows'] <= 3]

In [23]:
df3.shape

(15481, 38)

In [24]:
# Drop unwanated columns 
df3 = df3.drop(columns=['Unnamed: 0.1', 'Unnamed: 0', 'numPhotos', 'null_sum_rows'])

In [25]:
# Drop rows that are no longer needed
df3.drop(index=['corr_r', 'null_sum_col'], inplace = True)

In [26]:
df3.tail()

address_zipcode  bathrooms  bedrooms   latitude  livingArea  longitude  \
16477          78749.0        2.0       3.0  30.221336      2270.0 -97.879677   
16478          78749.0        2.0       4.0  30.216406      2645.0 -97.876045   
16479          78749.0        2.0       3.0  30.213816      1469.0 -97.873711   
16480          78749.0        2.5       3.0  30.218878      1700.0 -97.864960   
16481          78745.0        NaN       NaN  30.222555      3120.0 -97.805099   

       propertyTaxRate  resoFactsStats_bathrooms  \
16477             1.98                       2.0   
16478             1.98                       2.0   
16479             1.98                       2.0   
16480             1.98                       NaN   
16481             1.98                       0.0   

       resoFactsStats_bathroomsFull  resoFactsStats_bathroomsHalf  ...  \
16477                           0.0                           0.0  ...   
16478                           0.0                           0.0  ...   
16479                           2.0                           0.0  ...   
16480                           0.0                           0.0  ...   
16481                           0.0                           0.0  ...   

       schools_1_size  schools_1_studentsPerTeacher schools_1_totalCount  \
16477          1231.0                          17.0                  1.0   
16478          1275.0                          18.0                  1.0   
16479          1231.0                          17.0                  1.0   
16480          1231.0                          17.0                  1.0   
16481           660.0                          12.0                  1.0   

      schools_2_distance schools_2_rating  schools_2_studentsPerTeacher  \
16477                2.5              9.0                          18.0   
16478                2.1              9.0                          18.0   
16479                1.9              9.0                          18.0   
16480                7.0              7.0                          17.0   
16481                0.7              4.0                          14.0   

       schools_2_totalCount  latest_price  latest_saleyear  numPriceChanges  
16477                   1.0      425000.0           2018.0              1.0  
16478                   1.0      424999.0           2018.0              1.0  
16479                   1.0      316000.0           2018.0              5.0  
16480                   1.0      384900.0           2018.0              1.0  
16481                   1.0      515000.0           2018.0              1.0  

[5 rows x 34 columns]

In [27]:
df3['price_per_sqft'] = df3['latest_price'] / df3['livingArea']

In [28]:
df3.shape

(15479, 35)

In [29]:
# Evaluate the amount of missing values in each columns
df3.isnull().sum()

address_zipcode                    0
bathrooms                        346
bedrooms                         272
latitude                           0
livingArea                         0
longitude                          0
propertyTaxRate                    0
resoFactsStats_bathrooms         192
resoFactsStats_bathroomsFull     121
resoFactsStats_bathroomsHalf     121
resoFactsStats_bedrooms          223
garageSpaces                       0
hasGarage                          0
hasSpa                             0
hasView                            0
parking                            0
stories                          663
taxAnnualAmount                 1207
taxAssessedValue                 613
schools_0_distance                 0
schools_0_rating                   0
schools_0_size                     0
schools_0_studentsPerTeacher      91
schools_1_rating                   0
schools_1_size                     0
schools_1_studentsPerTeacher     371
schools_1_totalCount               0
s

In [30]:
# Evaluate the amount of missing values across rows
df3[df3.isnull().sum(axis = 1) > 0]

address_zipcode  bathrooms  bedrooms   latitude  livingArea  longitude  \
6              78732.0        NaN       3.0  30.329847      2326.0 -97.926910   
11             78732.0        NaN       NaN  30.330782      1480.0 -97.925789   
12             78732.0        1.0       1.0  30.330755       704.0 -97.926483   
21             78732.0        6.0       4.0  30.341303      4219.0 -97.919540   
57             78732.0        5.0       5.0  30.342978      4783.0 -97.922836   
...                ...        ...       ...        ...         ...        ...   
16466          78751.0        NaN       NaN  30.311029      3700.0 -97.719620   
16470          78723.0        2.0       3.0  30.317089      1665.0 -97.677162   
16474          78749.0        2.5       4.0  30.221970      2786.0 -97.875984   
16480          78749.0        2.5       3.0  30.218878      1700.0 -97.864960   
16481          78745.0        NaN       NaN  30.222555      3120.0 -97.805099   

       propertyTaxRate  resoFactsStats_bathrooms  \
6                 1.98                       0.0   
11                1.98                       0.0   
12                1.98                       1.0   
21                1.98                       6.0   
57                1.98                       5.0   
...                ...                       ...   
16466             1.98                       0.0   
16470             1.98                       2.0   
16474             1.98                       NaN   
16480             1.98                       NaN   
16481             1.98                       0.0   

       resoFactsStats_bathroomsFull  resoFactsStats_bathroomsHalf  ...  \
6                               0.0                           0.0  ...   
11                              0.0                           0.0  ...   
12                              1.0                           0.0  ...   
21                              4.0                           2.0  ...   
57                              5.0                           0.0  ...   
...                             ...                           ...  ...   
16466                           0.0                           0.0  ...   
16470                           2.0                           0.0  ...   
16474                           0.0                           0.0  ...   
16480                           0.0                           0.0  ...   
16481                           0.0                           0.0  ...   

       schools_1_studentsPerTeacher  schools_1_totalCount schools_2_distance  \
6                              17.0                   1.0                7.6   
11                             17.0                   1.0                7.5   
12                             17.0                   1.0                7.6   
21                             17.0                   1.0                6.9   
57                             17.0                   1.0                7.0   
...                             ...                   ...                ...   
16466                          16.0                   1.0                1.2   
16470                          13.0                   1.0                1.3   
16474                          17.0                   1.0                2.4   
16480                          17.0                   1.0                7.0   
16481                          12.0                   1.0                0.7   

      schools_2_rating schools_2_studentsPerTeacher  schools_2_totalCount  \
6                  7.0                         17.0                   1.0   
11                 7.0                         17.0                   1.0   
12                 7.0                         17.0                   1.0   
21                 7.0                         17.0                   1.0   
57                 7.0                         17.0                   1.0   
...                ...                          ...                   ...   
16466              8

In [31]:
df3.mean().round(1)

address_zipcode                  78735.7
bathrooms                            2.7
bedrooms                             3.4
latitude                            30.3
livingArea                        2150.0
longitude                          -97.8
propertyTaxRate                      2.0
resoFactsStats_bathrooms             2.6
resoFactsStats_bathroomsFull         2.0
resoFactsStats_bathroomsHalf         0.4
resoFactsStats_bedrooms              3.4
garageSpaces                         1.2
hasGarage                            0.6
hasSpa                               0.1
hasView                              0.2
parking                              1.2
stories                              1.5
taxAnnualAmount                   8800.7
taxAssessedValue                451189.5
schools_0_distance                   0.9
schools_0_rating                     6.1
schools_0_size                     636.7
schools_0_studentsPerTeacher        14.5
schools_1_rating                     5.1
schools_1_size  

In [32]:
df3.shape

(15479, 35)

In [33]:
(df3.address_zipcode.min(), df3.address_zipcode.max())

(78617.0, 78759.0)

In [34]:
df3.address_zipcode.dtype

dtype('float64')

In [35]:
df3['zipcode'] = df3.address_zipcode.astype('int')

In [36]:
df3 = df3.set_index(['zipcode'])

In [37]:
df3.head()

address_zipcode  bathrooms  bedrooms   latitude  livingArea  \
zipcode                                                                
78732            78732.0        5.0       4.0  30.354322      4060.0   
78732            78732.0        5.0       5.0  30.355553      4558.0   
78732            78732.0        4.0       4.0  30.352081      4749.0   
78732            78732.0        4.0       4.0  30.356226      4867.0   
78732            78732.0        4.0       5.0  30.341896      3485.0   

         longitude  propertyTaxRate  resoFactsStats_bathrooms  \
zipcode                                                         
78732   -97.911278             1.98                       5.0   
78732   -97.912544             1.98                       5.0   
78732   -97.912048             1.98                       4.0   
78732   -97.911697             1.98                       4.0   
78732   -97.907944             1.98                       4.0   

         resoFactsStats_bathroomsFull  resoFactsStats_bathroomsHalf  ...  \
zipcode                                                              ...   
78732                             4.0                           1.0  ...   
78732                             5.0                           0.0  ...   
78732                             3.0                           1.0  ...   
78732                             4.0                           0.0  ...   
78732                             3.0                           1.0  ...   

         schools_1_studentsPerTeacher  schools_1_totalCount  \
zipcode                                                       
78732                            17.0                   1.0   
78732                            17.0                   1.0   
78732                            17.0                   1.0   
78732                            17.0                   1.0   
78732                            17.0                   1.0   

        schools_2_distance schools_2_rating schools_2_studentsPerTeacher  \
zipcode                                                                    
78732                  6.0              7.0                         17.0   
78732                  6.1              7.0                         17.0   
78732                  6.1              7.0                         17.0   
78732                  6.0              7.0                         17.0   
78732                  6.2              7.0                         17.0   

         schools_2_totalCount  latest_price  latest_saleyear  numPriceChanges  \
zipcode                                                                         
78732                     1.0      715000.0           2018.0              1.0   
78732                     1.0     1025000.0           2020.0              2.0   
78732                     1.0      825000.0           2018.0              4.0   
78732                     1.0      849000.0           2019.0              7.0   
78732                     1.0      625000.0           2019.0              4.0   

         price_per_sqft  
zipcode                  
78732        176.108374  
78732        224.879333  
78732        173.720783  
78732        174.440107  
78732        179.340029  

[5 rows x 35 columns]

# load the US Income data for Austin and merge it with Austin House Listings cleaned data

In [38]:
# load the us_income data
df1_inc = pd.read_csv('../01_raw_data/us_income_zipcode.csv', low_memory=False)

In [39]:
df1_inc.shape

(364998, 111)

In [40]:
df2_inc = df1_inc[df1_inc['ZIP'].between(78617,78759)]

In [41]:
df2_inc.shape

(1025, 111)

In [42]:
list(df2_inc.columns)

['ZIP',
 'Geography',
 'Geographic Area Name',
 'Households',
 'Households Margin of Error',
 'Households Less Than $10,000',
 'Households Less Than $10,000 Margin of Error',
 'Households $10,000 to $14,999',
 'Households $10,000 to $14,999 Margin of Error',
 'Households $15,000 to $24,999',
 'Households $15,000 to $24,999 Margin of Error',
 'Households $25,000 to $34,999',
 'Households $25,000 to $34,999 Margin of Error',
 'Households $35,000 to $49,999',
 'Households $35,000 to $49,999 Margin of Error',
 'Households $50,000 to $74,999',
 'Households $50,000 to $74,999 Margin of Error',
 'Households $75,000 to $99,999',
 'Households $75,000 to $99,999 Margin of Error',
 'Households $100,000 to $149,999',
 'Households $100,000 to $149,999 Margin of Error',
 'Households $150,000 to $199,999',
 'Households $150,000 to $199,999 Margin of Error',
 'Households $200,000 or More',
 'Households $200,000 or More Margin of Error',
 'Households Median Income (Dollars)',
 'Households Median Income

In [43]:
df2_inc = df2_inc.set_index(['ZIP'])

In [44]:
df2_inc.head()

Geography Geographic Area Name  Households  \
ZIP                                                      
78617  860Z200US78617          ZCTA5 78617      7662.0   
78618  860Z200US78618          ZCTA5 78618       129.0   
78619  860Z200US78619          ZCTA5 78619      2159.0   
78620  860Z200US78620          ZCTA5 78620      6538.0   
78621  860Z200US78621          ZCTA5 78621      8103.0   

       Households Margin of Error  Households Less Than $10,000  \
ZIP                                                               
78617                       682.0                           3.0   
78618                        50.0                           6.2   
78619                       372.0                           0.0   
78620                       432.0                           1.5   
78621                       552.0                           5.0   

       Households Less Than $10,000 Margin of Error  \
ZIP                                                   
78617                                           2.3   
78618                                           6.8   
78619                                           2.2   
78620                                           1.0   
78621                                           2.3   

       Households $10,000 to $14,999  \
ZIP                                    
78617                            1.7   
78618                            0.8   
78619                            1.0   
78620                            3.4   
78621                            2.2   

       Households $10,000 to $14,999 Margin of Error  \
ZIP                                                    
78617                                            2.0   
78618                                            2.4   
78619                                            1.2   
78620                                            2.1   
78621                                            1.4   

       Households $15,000 to $24,999  \
ZIP                                    
78617                            3.7   
78618                            5.4   
78619                            0.0   
78620                            2.9   
78621                            2.4   

       Households $15,000 to $24,999 Margin of Error  ...  \
ZIP                                                   ...   
78617                                            1.8  ...   
78618                                            5.5  ...   
78619                                            2.2  ...   
78620                                            1.4  ...   
78621                                            1.4  ...   

       Nonfamily Households $150,000 to $199,999  \
ZIP                                                
78617                                        2.1   
78618                                       14.8   
78619                                        0.0   
78620                                        4.8   
78621                                       12.2   

       Nonfamily Households $150,000 to $199,999 Margin of Error  \
ZIP                                                                
78617                                                3.4           
78618                                               22.7           
78619                                               25.9           
78620                                                3.9           
78621                                               15.1           

       Nonfamily Households $200,000 or More  \
ZIP                                            
78617                                    2.6   
78618                                   14.8   
78619                                    0.0   
78620                                    5.4   
78621                                    2.0   

       Nonfamily Households $200,000 or More Margin of Error  \
ZIP                                                            
78617                                                4.0   

In [45]:
#Analyze Data across a row
dict(df2_inc.iloc[0])

{'Geography': '860Z200US78617',
 'Geographic Area Name': 'ZCTA5 78617',
 'Households': 7662.0,
 'Households Margin of Error': 682.0,
 'Households Less Than $10,000': 3.0,
 'Households Less Than $10,000 Margin of Error': 2.3,
 'Households $10,000 to $14,999': 1.7,
 'Households $10,000 to $14,999 Margin of Error': 2.0,
 'Households $15,000 to $24,999': 3.7,
 'Households $15,000 to $24,999 Margin of Error': 1.8,
 'Households $25,000 to $34,999': 6.6,
 'Households $25,000 to $34,999 Margin of Error': 2.4,
 'Households $35,000 to $49,999': 19.2,
 'Households $35,000 to $49,999 Margin of Error': 6.4,
 'Households $50,000 to $74,999': 20.1,
 'Households $50,000 to $74,999 Margin of Error': 4.9,
 'Households $75,000 to $99,999': 10.3,
 'Households $75,000 to $99,999 Margin of Error': 3.9,
 'Households $100,000 to $149,999': 23.0,
 'Households $100,000 to $149,999 Margin of Error': 5.4,
 'Households $150,000 to $199,999': 6.0,
 'Households $150,000 to $199,999 Margin of Error': 2.8,
 'Household

In [46]:
selected_cols = [col for col in df2_inc.columns if 'Mean Income' in col and 'Margin of Error' not in col]

In [47]:
df2_inc = df2_inc[selected_cols]

In [48]:
df2_inc.head()

Households Mean Income (Dollars)  Families Mean Income (Dollars)  \
ZIP                                                                       
78617                           99354.0                        106461.0   
78618                          105468.0                         92582.0   
78619                          165887.0                        169968.0   
78620                          151574.0                        165526.0   
78621                           94097.0                         95930.0   

       Married-Couple Families Mean Income (Dollars)  \
ZIP                                                    
78617                                            NaN   
78618                                            NaN   
78619                                            NaN   
78620                                            NaN   
78621                                            NaN   

       Nonfamily Households Mean Income (Dollars)  
ZIP                                                
78617                                     56019.0  
78618                                    154152.0  
78619                                     75649.0  
78620                                     70972.0  
78621                                     80298.0

In [49]:
df2_inc.shape

(1025, 4)

In [50]:
df2_inc_average = df2_inc.mean(axis = 1).round(1)

In [51]:
merged_df = pd.merge(df3, df2_inc_average, on='zipcode', how='left')

ValueError: Cannot merge a Series without a name

In [ ]:
# Sample DataFrame
df = pd.DataFrame({'Values': [123, 4567, 89012, 345678]})

# Pad zeros to the left of the values to ensure they have 5 digits
df['Formatted_Values'] = df['Values'].astype(str).str.zfill(5)

# Print the DataFrame
print("Original DataFrame:")
print(df)